scipio/src/channels/local_channel.rs

```diff
@@ -0,0 +1,586 @@
+///
+/// # Examples
+///
+/// ```
+/// use scipio::{LocalExecutor, Local, Latency, Shares};
+/// use scipio::channels::LocalChannel;
+/// use futures::stream::StreamExt;
+///
+/// let ex = LocalExecutor::make_default();
+/// ex.run(async move {
+///     let task_queue =
+///         Local::create_task_queue(
+///             Shares::default(), Latency::NotImportant, "example");
+///     
+///     let (sender, mut receiver) = LocalChannel::new_unbounded();
+///     let h = Local::local_into(async move {
+///         assert_eq!(receiver.next().await.unwrap(), 0);
+///     }, task_queue).unwrap().detach();
+///     sender.send(0);
+///     drop(sender);
+///     h.await;
+/// });
+/// ```
+/// [`TaskQueue`]: struct.TaskQueue.html
+/// [`LocalSender`]: struct.LocalSender.html
+/// [`LocalReceiver`]: struct.LocalReceiver.html
+/// [`push`]: struct.LocalSender.html#method.push
+/// [`None`]: https://doc.rust-lang.org/std/option/enum.Option.html#variant.None
+/// [`BrokenPipe`]: https://doc.rust-lang.org/std/io/enum.ErrorKind.html#variant.BrokenPipe
+/// [`channel`]: https://doc.rust-lang.org/std/sync/mpsc/fn.channel.html
+/// [`next`]: https://docs.rs/futures/0.3.6/futures/stream/trait.StreamExt.html#method.next
+/// [`StreamExt`]: https://docs.rs/futures/0.3.6/futures/stream/trait.StreamExt.html
+pub struct LocalChannel<T> {
+    state: Rc<RefCell<State<T>>>,
+}
+
+impl<T> LocalChannel<T> {
+    fn new(capacity: ChannelCapacity) -> (LocalSender<T>, LocalReceiver<T>) {
+        let channel = match capacity {
+            ChannelCapacity::Unbounded => VecDeque::new(),
+            ChannelCapacity::Bounded(x) => VecDeque::with_capacity(x),
+        };
+
+        let state = Rc::new(RefCell::new(State {
+            capacity,
+            channel,
+            send_waiters: Vec::new(),
+            recv_waiters: Vec::new(),
+            receiver_alive: true,
+            sender_alive: true,
+        }));
+
+        (
+            LocalSender {
+                channel: LocalChannel {
+                    state: state.clone(),
+                },
+            },
+            LocalReceiver {
+                channel: LocalChannel {
+                    state: state.clone(),
+                },
+            },
+        )
+    }
+
+    fn push(&self, item: T) -> io::Result<()> {
+        let mut state = self.state.borrow_mut();
+        if !state.receiver_alive {
+            return Err(io::Error::new(io::ErrorKind::BrokenPipe, ""));
+        }
+
+        if let ChannelCapacity::Bounded(x) = state.capacity {
+            if state.channel.len() >= x {
+                return Err(io::Error::new(io::ErrorKind::Other, "Channel Full"));
+            }
+        }
+
+        state.channel.push_back(item);
+        if !state.recv_waiters.is_empty() {
+            let waiters = std::mem::replace(&mut state.recv_waiters, Vec::new());
+            drop(state);
+
+            for w in waiters {
+                w.wake();
+            }
```

Comment:
Shouldn't it wake just one?

Reply:
That's a very good question. It's clearly not wrong to wake them all since they'll just be put to sleep.
Intuitively it feels to me that if we are pushing and consuming a single item, a single notification would be enough and you are right.

I need to convince myself that a lost notification is not possible in the parallel executors, though.

Reply:
Not totally clear on what the parallel executors are, but in general, a _check - signal sleep - redo check - sleep_ pattern works well when there's parallelism.

Reply:
Actually I think we should be waking just one here. Otherwise imagine a blocked channel with 1,000,000 waiters that gets 1,000,000 elements pushed: it will call `wake` a lot as it becomes quadratic.